src/components/input/input.js

```diff
@@ -404,6 +404,9 @@ function mdMaxlengthDirective($animate) {
       if (!angular.isNumber(maxlength) || maxlength < 0) {
         return true;
       }
+      if (angular.isNumber(modelValue)) {
+        modelValue = modelValue.toString();
```

Comment:
That would be ideal. [It has been mentioned before](https://github.com/angular/material/issues/1960#issuecomment-83494622) as a possible solution, so perhaps line `410` can just be changed to `return ( element.val() || viewValue || '' ).length <= maxlength;`